src/amr/data/field/coarsening/magnetic_field_coarsener.hpp

```diff
@@ -0,0 +1,139 @@
+#ifndef PHARE_MAGNETIC_FIELD_COARSENER
+#define PHARE_MAGNETIC_FIELD_COARSENER
+
+#include "core/data/grid/gridlayoutdefs.hpp"
+#include "core/hybrid/hybrid_quantities.hpp"
+#include "core/utilities/constants.hpp"
+
+
+#include <SAMRAI/hier/Box.h>
+#include <stdexcept>
+
+namespace PHARE::amr
+{
+using core::dirX;
+using core::dirY;
+using core::dirZ;
+/** @brief This class gives an operator() that performs the coarsening of N fine nodes onto a
+ * given coarse node
+ *
+ * A MagneticFieldCoarsener object is created each time the refine() method of the
+ * FieldCoarsenOperator is called and its operator() is called for each coarse index.
+ * It is the default coarsening policy and used for any field that does not come with
+ * specific constraints (such as conserving some property in the coarsening process).
+ *
+ *
+ * This coarsening operation is defined so to conserve the magnetic flux.
+ * This is done by assigning to a magnetic field component on a coarse face, the average
+ * of the enclosed fine faces
+ *
+ */
+template<std::size_t dimension>
+class MagneticFieldCoarsener
+{
+public:
+    MagneticFieldCoarsener(std::array<core::QtyCentering, dimension> const centering,
+                           SAMRAI::hier::Box const& sourceBox,
+                           SAMRAI::hier::Box const& destinationBox,
+                           SAMRAI::hier::IntVector const& ratio)
```

Comment:
are you using `SAMRAI::hier::IntVector const& ratio` cause it was taken from a code sample?
do we need it considering we only support `ratio==2` (assuming it's the same thing)

Reply:
I'm taking the IntVector because that's what the `coarsen()` function gives it, that function comes from SAMRAI and I didn't want to convert it for no good reason. You may wonder why this function takes this argument even though `ratio_` is constexpr set to 2 and that argument is not even used. But that class is used as a template param in `coarsen` called `FieldCoarsenerPolicy` and the other "policy" , namely `DefaultFieldCoarsener` uses this ratio....